io/LasWriter.cpp

```diff
@@ -581,14 +583,55 @@ void LasWriter::addVlr(const std::string& userId, uint16_t recordId,
 void LasWriter::addVlr(const std::string& userId, uint16_t recordId,
    const std::string& description, std::vector<char>&& data)
 {
-    addVlr(las::Evlr(userId, recordId, description, data));
+    las::Evlr vlr(userId, recordId, description, data);
+    addVlr(vlr);
 }
 
+    class Predicate
+    {
+    public:
+        Predicate(const std::string& name) : m_name(name)
+        {}
+
+        bool operator()(MetadataNode m)
+            { return m.name() == m_name; }
+    private:
+        std::string m_name;
+    };
 
-/// Add a standard or variable-length VLR depending on the data size.
-/// \param  evlr  VLR to add.
-void LasWriter::addVlr(const las::Evlr& evlr)
+void LasWriter::addVlr(las::Evlr& evlr, MetadataNode m)
 {
+    auto pred = [evlr](MetadataNode n)
+        { return Utils::iequals(n.name(),  evlr.metadataId); };
+
+    auto setVlrDataFromMetadata = [m, pred, this](las::Evlr& v)
+    {
+        // Check if the vlr has a metadataId set
+        // if so, go find it in our metadata, copy it,
+        // and then wipe the metadataId from the eVLR because
+        // we cannot ever find/set that again
+        if (v.metadataId.size() > 0)
+        {
+            MetadataNode node = m.find(pred);
+
+            if (node.valid())
+            {
+                // Yuck
+                std::string s = node.value();
+                v.dataVec.insert(v.dataVec.end(), (char *)s.data(), (char *)(s.data() + s.size()));
+
+                // Wipe off our metadataId now that we have
+                // set the dataVect to it
+                v.metadataId = std::string("");
```

Comment:
v.metadataId.clear();

seems cleaner, though I don't think you need to do it at all.